File: bindings/lua/insight/signal/filter_design.lua
--- Filter design functions.
-- Provides FIR filter design, Kaiser window parameter estimation,
-- and complex sorting utilities.
-- @module insight.signal.filter_design

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

--- Compute the Kaiser beta parameter for a given attenuation.
-- @number attenuation Stopband attenuation in dB.
-- @number Kaiser beta parameter.
M.kaiser_beta = _wrap({ "attenuation" }, function(attenuation)
  return sig.kaiser_beta(attenuation)
end)

--- Compute the Kaiser window attenuation for given parameters.
-- @int n Number of taps.
-- @number width Transition width (normalized).
-- @number Attenuation in dB.
M.kaiser_atten = _wrap({ "n", "width" }, function(n, width)
  return sig.kaiser_atten(n, width)
end)

--- Design a FIR filter using the window method.
-- @int numtaps Number of filter taps.
-- @number cutoff Cutoff frequency (normalized, 0 to 1).
-- @string[opt="lowpass"] type Filter type: 'lowpass', 'highpass', 'bandpass', 'bandstop'.
-- @bool[opt=false] pass_zero If true, the DC gain is nonzero at frequency 0.
-- @treturn Array FIR filter coefficients.
M.firwin = _wrap({ "numtaps", "cutoff", "type", "pass_zero" }, function(numtaps, cutoff, type, pass_zero)
  return sig.firwin(numtaps, cutoff, type or "lowpass", pass_zero or false)
end)

--- Design a FIR filter using frequency sampling.
-- @int numtaps Number of filter taps.
-- @tparam table freq Frequency breakpoints (normalized, 0 to 1).
-- @tparam table gain Desired gain at each frequency breakpoint.
-- @treturn Array FIR filter coefficients.
M.firwin2 = _wrap({ "numtaps", "freq", "gain" }, function(numtaps, freq, gain)
  return sig.firwin2(numtaps, freq, gain)
end)

--- Sort complex numbers by real part, then imaginary part.
-- @tparam Array x Complex input array.
-- @treturn Array Sorted complex array.
M.cmplx_sort = _wrap({ "x" }, function(x)
  return sig.cmplx_sort(x)
end)

return M

File: bindings/python/insight/signal/filter_design.py
"""Filter design functions for signal processing.

Provides FIR filter design (window method), Kaiser window parameter
estimation, and helper utilities for IIR filter design.
"""

from insight._insight import signal as _signal

__all__ = [
    "kaiser_beta",
    "kaiser_atten",
    "firwin",
    "firwin2",
    "cmplx_sort",
]


def kaiser_beta(a):
    """Compute the Kaiser window beta parameter for a given stopband attenuation.

    Args:
        a: Stopband attenuation in dB (positive value).

    Returns:
        Kaiser window beta parameter.
    """
    return _signal.kaiser_beta(a)


def kaiser_atten(numtaps, width):
    """Compute the Kaiser window stopband attenuation.

    Args:
        numtaps: Number of filter taps (length of the FIR filter).
        width: Transition width as a fraction of the Nyquist frequency.

    Returns:
        Stopband attenuation in dB.
    """
    return _signal.kaiser_atten(numtaps, width)


def firwin(numtaps, cutoff, window="hamming", pass_zero="lowpass", scale=True):
    """Design a FIR filter using the window method.

    Args:
        numtaps: Number of taps in the FIR filter (filter length).
        cutoff: Cutoff frequency or array of cutoff frequencies.
            For a lowpass or highpass filter, a single cutoff is required.
            For bandpass or bandstop, two cutoffs are required.
        window: Window function to use.  Default is ``'hamming'``.
        pass_zero: Filter type.  One of ``'lowpass'``, ``'highpass'``,
            ``'bandpass'``, ``'bandstop'``.  Default is ``'lowpass'``.
        scale: Whether to scale the filter so that the frequency response
            at the first (or only) passband is unity.  Default is True.

    Returns:
        FIR filter coefficients as a 1-D array.
    """
    return _signal.firwin(numtaps, cutoff, window=window, pass_zero=pass_zero, scale=scale)


def firwin2(numtaps, freq, gain, nfreqs=None, window="hamming", antisymmetric=False):
    """Design a FIR filter with arbitrary frequency response.

    Args:
        numtaps: Number of taps in the FIR filter.
        freq: Sequence of frequency breakpoints (normalized to [0, 1] where
            1 is the Nyquist frequency).
        gain: Gain at each frequency breakpoint (same length as *freq*).
        nfreqs: Number of interpolation points for the frequency response.
            If ``None``, uses a default value chosen by the backend.
        window: Window function to use.  Default is ``'hamming'``.
        antisymmetric: If True, design a type III or IV filter with
            antisymmetric impulse response.  Default is False.

    Returns:
        FIR filter coefficients as a 1-D array.
    """
    kwargs = {"window": window, "antisymmetric": antisymmetric}
    if nfreqs is not None:
        kwargs["nfreqs"] = nfreqs
    return _signal.firwin2(numtaps, freq, gain, **kwargs)


def cmplx_sort(p):
    """Sort complex numbers by real part, then by imaginary part.

    Args:
        p: Sequence of complex numbers (array-like).

    Returns:
        A tuple ``(sorted_p, index)`` where *sorted_p* is the sorted array
        and *index* maps from sorted to original order.
    """
    return _signal.cmplx_sort(p)
